src/main/java/net/fabricmc/loader/discovery/ModResolver.java

```diff
@@ -87,17 +87,80 @@ private static IVecInt toVecInt(IntStream stream) {
 		return new VecInt(stream.toArray());
 	}
 
-	private boolean matches(ModDependency dependency, Map<String, ModCandidate> result) {
-		if (!result.containsKey(dependency.getModId())) {
+	/** @param errorList The list of errors. The returned list of errors all need to be prefixed with "it " in order to make sense. */
+	private static boolean isModIdValid(String modId, List<String> errorList) {
+		// A more useful error list for MOD_ID_PATTERN
+		if (modId.isEmpty()) {
+			errorList.add("is empty!");
 			return false;
 		}
-
-		return dependency.matches(result.get(dependency.getModId()).getInfo().getVersion());
+		if (modId.length() == 1) {
+			errorList.add("is only a single character! (It must be at least 2 characters long)!");
+		} else if (modId.length() > 64) {
+			errorList.add("has more than 64 characters!");
+		}
+		char first = modId.charAt(0);
+		if (first < 'a' || first > 'z') {
+			errorList.add("starts with an invalid character '" + first + "' (it must be a lowercase a-z)");
```

Comment:
The mod id must be lowercase throughout, this could be changed to make that clear as this is the most common mistake someone may make.

Reply:
That's true, although the only concern I have is that the rules are different for the first letter and the rest of the ID - I don't want to imply that 0-9 and `-`, `_` aren't valid characters for the rest of it.

Reply:
Ok, I've changed it from:

`"starts with an invalid character '" + first + "' (it must be a lowercase a-z)"`

to:

`"starts with an invalid character '" + first + "' (it must be a lowercase a-z - upper case isn't allowed anywhere in the ID)"`

is that reasonable?